File: Leetcode 3801-3900/3812/3812.py
# Leetcode 3812: Minimum Edges Toggles on a Tree
# https://leetcode.com/problems/minimum-edge-toggles-on-a-tree/
# Solved on 22nd of January, 2026
class Solution:
    def minimumFlips(self, n: int, edges: list[list[int]], start: str, target: str) -> list[int]:
        """
        Calculates the minimum set of edge indices to toggle to transform start states to target states.

        Args:
            n (int): Number of nodes in the tree.
            edges (list[list[int]]): List of edges [u, v].
            start (str): Initial binary states of nodes.
            target (str): Target binary states of nodes.

        Returns:
            list[int]: Sorted list of edge indices to toggle, or [-1] if impossible.
        """
        needsFlip = [0] * n
        mismatchCount = 0
        for i in range(n):
            if start[i] != target[i]:
                needsFlip[i] = 1
                mismatchCount += 1

        if mismatchCount % 2 != 0:
            return [-1]

        adjList = [[] for _ in range(n)]
        for index, (u, v) in enumerate(edges):
            adjList[u].append((v, index))
            adjList[v].append((u, index))

        bfsQueue = [0]
        visitOrder = []
        visited = [False] * n
        visited[0] = True
        parentMap = {}

        head = 0
        while head < len(bfsQueue):
            u = bfsQueue[head]
            head += 1
            visitOrder.append(u)

            for v, edgeIndex in adjList[u]:
                if not visited[v]:
                    visited[v] = True
                    parentMap[v] = (u, edgeIndex)
                    bfsQueue.append(v)

        resultEdges = []
        for i in range(n - 1, -1, -1):
            u = visitOrder[i]
            if u == 0:
                continue

            if needsFlip[u] == 1:
                parent, edgeIndex = parentMap[u]
                resultEdges.append(edgeIndex)
                needsFlip[parent] ^= 1

        resultEdges.sort()
        return resultEdges

File: Leetcode 3801-3900/3812/3812.java
// Leetcode 3812: Minimum Edge Toggles on a Tree
// https://leetcode.com/problems/minimum-edge-toggles-on-a-tree/
// Solved on 22nd of January, 2026
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

class Solution {
    private List<Integer> result;
    private List<Integer>[] graph;
    private int[] diff;
    private int[][] edgeList;

    /**
     * Finds the minimum number of edge toggles required to transform the start state
     * of nodes to the target state in a tree. Each toggle on an edge flips the state
     * of both connected nodes.
     * @param n The number of nodes in the tree.
     * @param edges A 2D array where edges[i] = [u, v] represents an edge between nodes u and v.
     * @param start A binary string representing the initial state of each node.
     * @param target A binary string representing the target state of each node.
     * @return A list of indices of the edges that need to be toggled, or a list containing -1 if impossible.
     */
    public List<Integer> minimumFlips(int n, int[][] edges, String start, String target) {
        diff = new int[n];
        char[] s = start.toCharArray();
        char[] t = target.toCharArray();
        for (int i = 0; i < n; i++) {
            diff[i] = s[i] ^ t[i];
        }

        graph = new ArrayList[n];
        for (int i = 0; i < n; i++) {
            graph[i] = new ArrayList<>();
        }
        for (int i = 0; i < edges.length; i++) {
            graph[edges[i][0]].add(i);
            graph[edges[i][1]].add(i);
        }

        edgeList = edges;
        result = new ArrayList<>();

        dfs(0, -1);

        if (diff[0] != 0) {
            List<Integer> impossible = new ArrayList<>();
            impossible.add(-1);
            return impossible;
        }

        Collections.sort(result);
        return result;
    }

    private void dfs(int u, int p) {
        for (int idx : graph[u]) {
            int v = (edgeList[idx][0] == u) ? edgeList[idx][1] : edgeList[idx][0];
            if (v != p) {
                dfs(v, u);
                if (diff[v] == 1) {
                    result.add(idx);
                    diff[v] ^= 1;
                    diff[u] ^= 1;
                }
            }
        }
    }
}